### PR TITLE
Rename 'level' to 'subsampling_divisions' with deprecation shims

### PR DESCRIPTION
**Pull request by blooop**
## Summary
- Renames the public sampling knob from `level` to `subsampling_divisions` across the entire API (parameters, constants, helpers, docs, examples, tests)
- Introduces `SUBSAMPLING_DIVISIONS_SAMPLES`, `with_subsampling_divisions()`, `subsampling_divisions_to_samples()` as new primitives
- Adds backward-compatible deprecation shims so existing `level`-based code continues to work with warnings
- Adds `UNSET` sentinel to distinguish "not provided" from "explicitly passed the default" in conflict detection (addresses reviewer feedback about magic default `2`)
- Adds conflict checks: passing both `level` and `subsampling_divisions` (or `max_level` and `max_subsampling_divisions`) raises `TypeError`
- Extracts shared `normalize_subsampling_divisions_kwargs` helper to centralize deprecation logic (addresses reviewer feedback about duplication)

## Test plan
- [x] All 1382 existing tests pass (0 failures)
- [x] New sentinel conflict tests: `run(subsampling_divisions=2, level=3)` correctly raises even when value matches default
- [x] Deprecation path tests for `level`, `max_level`, `min_level`, `LEVEL_SAMPLES`, `with_level`
- [x] `pixi run ci` passes (format, lint, test with coverage)
- [x] `pixi run generate-docs` succeeds

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bencher/bench_runner.py [hunk 1]
```diff
@@ -134,28 +134,37 @@ def _generate_name(self) -> str:
     @staticmethod
     def setup_run_cfg(
         run_cfg: BenchRunCfg | None = None,
-        level: int = 2,
+        fidelity: int = 2,
         cache_samples: bool = False,
         over_time: bool | None = None,
+        level: int | None = None,
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/bench_runner.py]: Addressed: `setup_run_cfg` now uses the `UNSET` sentinel instead of comparing against the magic default `2`. Passing both `level` and `subsampling_divisions` (even when subsampling_divisions equals the default) raises `TypeError`. See `test_setup_run_cfg_level_and_subsampling_divisions_default_conflict_raises` for the new edge-case test.

File: bencher/variables/inputs.py [hunk 2]
```diff
@@ -698,13 +700,22 @@ def sweep(
...
+    if max_level is not None:
+        warnings.warn(
+            "The 'max_level' parameter is deprecated; use 'max_fidelity' instead.",
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        if max_fidelity is None:
+            max_fidelity = max_level
+
+    if max_fidelity is not None and max_fidelity <= 0:
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/variables/inputs.py]: Addressed: `sweep()` now raises `TypeError` when both `max_level` and `max_subsampling_divisions` are provided, matching the behavior of `BenchRunner.run` and `bn.run`. Test: `test_max_level_and_max_subsampling_divisions_conflict_raises`.

File: bencher/variables/inputs.py [hunk 3]
```diff
@@ -745,29 +756,37 @@ def p(
...
+    """Apply fidelity-based sampling to a list of values.
 
-    Uses the same level→sample-count table as SweepBase.with_level and picks
+    Uses the same fidelity→sample-count table as SweepBase.with_fidelity and picks
     evenly spaced items from *arr* by index.
 
     Args:
         arr (list): list of values to sample from
-        level (int): The sampling level to apply (higher levels provide more samples)
+        fidelity (int): The sampling fidelity to apply (higher fidelity provides more samples)
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/variables/inputs.py]: Addressed: `with_subsampling_divisions()` now derives the cap from `len(SUBSAMPLING_DIVISIONS_SAMPLES) - 1` instead of hard-coding `12`, so it stays aligned if the table grows.

File: test/test_usability.py [hunk 4]
```diff
@@ -18,35 +18,51 @@ def benchmark(self):
...
+class TestFidelitySamples(unittest.TestCase):
     def test_exported_from_package(self):
-        self.assertIs(bn.LEVEL_SAMPLES, LEVEL_SAMPLES)
+        self.assertIs(bn.FIDELITY_SAMPLES, FIDELITY_SAMPLES)
 
     def test_first_entry_is_zero(self):
-        self.assertEqual(LEVEL_SAMPLES[0], 0)
+        self.assertEqual(FIDELITY_SAMPLES[0], 0)
 
     def test_monotonically_increasing(self):
```
[the remark this answers is not in the document]
Comment (blooop) [reply test/test_usability.py:31]: Addressed: `TestWithLevelBackwardCompat.test_with_level_alias_emits_warning_and_delegates` verifies `bn.with_level` emits `DeprecationWarning`, is callable, and produces the same result as `with_subsampling_divisions`.

File: test/test_bench_runner.py [hunk 5]
```diff
@@ -183,6 +183,71 @@ def simple_test(run_cfg: bn.BenchRunCfg, report: bn.BenchReport) -> bn.BenchCfg:
...
+            )
+
+    def test_benchrunner_run_level_and_fidelity_conflict_raises(self):
+        """Passing both level= and fidelity= to BenchRunner.run() raises TypeError."""
+        br = bn.BenchRunner("conflict_test")
+        br.add(lambda run_cfg, report: bn.BenchCfg())
+        with self.assertRaises(TypeError, msg="Cannot pass both"):
+            br.run(fidelity=3, level=4)
+
     def test_benchrunner_no_name_instantiation(self):
```
[the remark this answers is not in the document]
Comment (blooop) [reply test/test_bench_runner.py]: Addressed: Added both tests — `test_benchrunner_run_max_level_and_max_subsampling_divisions_conflict_raises` (raises `TypeError`) and `test_benchrunner_run_max_level_deprecation_translates` (warns and maps to `max_subsampling_divisions`).

File: test/test_sweep_helper.py [hunk 6]
```diff
@@ -19,7 +19,7 @@ def test_returns_correct_dict_name_only(self):
             "name": "param1",
             "values": None,
             "samples": None,
-            "max_level": None,
+            "max_fidelity": None,
```
[the remark this answers is not in the document]
Comment (blooop) [reply test/test_sweep_helper.py]: Addressed: Added `test_max_level_deprecated_alias_maps_to_max_subsampling_divisions` (warns and maps) and `test_max_level_and_max_subsampling_divisions_conflict_raises` (raises `TypeError` when both provided).

File: bencher/bench_runner.py [hunk 7]
```diff
@@ -279,10 +288,10 @@ def _execute_bench_fn(
 
     def run(
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/bench_runner.py:296]: Addressed: Extracted `normalize_subsampling_divisions_kwargs()` in `bencher/utils.py` as the single shared helper for all `level`/`max_level` deprecation logic. Both `BenchRunner.run()` and `bn.run()` delegate to it. `setup_run_cfg` uses the same `UNSET` sentinel pattern. The magic default `2` issue is also resolved — an explicit `UNSET` sentinel distinguishes "not provided" from "provided the default value".

File: bencher/run.py [hunk 8]
```diff
@@ -138,6 +139,27 @@ def run(
     """
     from bencher.bench_runner import BenchRunner, _resolve_cache_samples
 
+    if "level" in kwargs:
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/run.py]: Addressed: `normalize_subsampling_divisions_kwargs()` in `bencher/utils.py` is the shared helper used by both `bn.run()` and `BenchRunner.run()`. Legacy `level`/`max_level` handling is fully centralized — neither function body mentions them directly.

File: bencher/__init__.py [hunk 9]
```diff
@@ -155,3 +155,23 @@
 from .class_enum import ClassEnum, ExampleEnum
 from .factories import create_bench, create_bench_runner
 from .run import run
+
+
+def __getattr__(name: str):
```
[the remark this answers is not in the document]
Comment (blooop) [reply bencher/__init__.py:166]: Addressed: `__getattr__` now uses a data-driven `_DEPRECATED_ALIASES` map with dynamic forwarding via `sys.modules[__name__]`. No hard-coded conditionals or nested imports.

**Comment by blooop** [comment]
All 9 review comments addressed. Summary of the two high-level points:

1. **Magic default `2` → `UNSET` sentinel**: `normalize_subsampling_divisions_kwargs()`, `BenchRunner.run()`, `bn.run()`, and `setup_run_cfg()` now use an explicit `UNSET` sentinel as the default for `subsampling_divisions`. This means `run(subsampling_divisions=2, level=3)` correctly raises `TypeError` instead of silently preferring `level`. New tests cover this edge case.

2. **Centralized deprecation handling**: All `level`/`max_level` → `subsampling_divisions`/`max_subsampling_divisions` logic lives in the shared `normalize_subsampling_divisions_kwargs()` helper in `bencher/utils.py`. `BenchRunCfg.__init__` and `with_defaults()` handle their own `level` param directly (since they use `**params`/`**defaults` dict patterns), but the deprecation message and conflict behavior is consistent everywhere.
